app.py

```diff
@@ -0,0 +1,125 @@
+from sklearn.feature_extraction.text import TfidfTransformer
+
+# Import necessary function(s) from tfidf.py
+from tfidf import get_keywords
+
+class TfidfKeywordextractor(ClamsApp):
+
+    def __init__(self):
+        super().__init__()
+
+    def _appmetadata(self):
+        pass
+
+    def _annotate(self, mmif: Mmif, **parameters) -> Mmif:
+        self.mmif = mmif if type(mmif) is Mmif else Mmif(mmif)
+        idf_feature_file = parameters['idfFeatureFile']
+        topn = parameters['topN']
+
+        text_slicer_checker = self.mmif.get_view_contains(DocumentTypes.TextDocument)
+        new_view = self._new_view(parameters)
+        if text_slicer_checker is None:
+            # scenario 1: single text document input.
+            for doc in self.mmif.get_documents_by_type(DocumentTypes.TextDocument):
+                self._keyword_extractor(doc, new_view, doc.long_id, idf_feature_file, topn)
+        else:
+            # scenario 2: input document is the one generated from the text slicer.
+            docs = text_slicer_checker.get_annotations(DocumentTypes.TextDocument)
+            for doc in docs:
+                text = doc.text_value
+                if len(text.split()) > topn:
+                    self._keyword_extractor(doc, new_view, doc.long_id, idf_feature_file, topn)
+
+        # return the MMIF object
+        return self.mmif
+
+    def _new_view(self, runtime_config):
+        view = self.mmif.new_view()
+        view.metadata.app = self.metadata.identifier
+        self.sign_view(view, runtime_config)
+        view.new_contain(DocumentTypes.TextDocument)
+        view.new_contain(AnnotationTypes.Alignment)
+        return view
+
+    def _keyword_extractor(self, doc, new_view, full_doc_id, idf_feature_file, topn):
+        """Run the keyword extractor over the document and add annotations to the view, using the
+        full document identifier (which may include a view identifier) for the document
+        property."""
+        text = doc.text_value
+
+        # load idf values
+        with open(idf_feature_file, 'rb') as f:
+            idf_feature_values = pickle.load(f)
+            idf_values = idf_feature_values["idf_values"]
+            feature_dict = idf_feature_values["feature_dict"]
+
+        # restore the values
+        cv = CountVectorizer(vocabulary=feature_dict)
+        tfidf_transformer = TfidfTransformer()
+        tfidf_transformer.idf_ = idf_values
+
+        # get keywords for the document
+        keywords_dict = get_keywords(text, feature_dict, topn, tfidf_transformer, cv)
+        keywords = ""
+        tfidf_values = ""
+        for keyword, tfidf_value in keywords_dict.items():
+            keywords += keyword + " "
+            tfidf_values += str(tfidf_value) + " "
+
+        # create the document to store the keywords
+        keywords_doc = new_view.new_textdocument(text=keywords.strip())
+        keywords_doc.add_property('scores', tfidf_values.strip())
```

Comment:
I'm not sure we need to cast scores to a single string, don't see any obvious reason that helps any downstream task. Can we use a list of real numbers for the value for this?